File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/futoshiki/domain/FutoshikiGame.java
package com.zwstudio.logicpuzzlesandroid.puzzles.futoshiki.domain;

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame;
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fj.F2;

public class FutoshikiGame extends CellsGame<FutoshikiGame, FutoshikiGameMove, FutoshikiGameState> {
    public static Position offset[] = {
            new Position(-1, 0),
            new Position(0, 1),
            new Position(1, 0),
            new Position(0, -1),
    };

    public char[] objArray;
    public Map<Position, Character> pos2hint = new HashMap<>();

    public char get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public char get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, char obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, char obj) {
        set(p.row, p.col, obj);
    }

    public FutoshikiGame(List<String> layout, GameInterface<FutoshikiGame, FutoshikiGameMove, FutoshikiGameState> gi, GameDocumentInterface gdi) {
        super(gi, gdi);
        size = new Position(layout.size(), layout.get(0).length());
        objArray = new char[rows() * cols()];
        for (int r = 0; r < rows(); r++) {
            String str = layout.get(r);
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                char ch = str.charAt(c);
                set(p, ch);
                if ((r % 2 != 0 || c % 2 != 0) && ch != ' ')
                    pos2hint.put(p, ch);
            }
        }

        FutoshikiGameState state = new FutoshikiGameState(this);
        states.add(state);
        levelInitilized(state);
    }

    private boolean changeObject(FutoshikiGameMove move, F2<FutoshikiGameState, FutoshikiGameMove, Boolean> f) {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size()).clear();
            moves.subList(stateIndex, states.size()).clear();
        }
        FutoshikiGameState state = cloner.deepClone(state());
        boolean changed = f.f(state, move);
        if (changed) {
            states.add(state);
            stateIndex++;
            moves.add(move);
            moveAdded(move);
            levelUpdated(states.get(stateIndex - 1), state);
        }
        return changed;
   }

    public boolean switchObject(FutoshikiGameMove move) {
        return changeObject(move, (state, move2) -> state.switchObject(move2));
    }

    public boolean setObject(FutoshikiGameMove move) {
        return changeObject(move, (state, move2) -> state.setObject(move2));
    }

    public char getObject(Position p) {
        return state().get(p);
    }

    public char getObject(int row, int col) {
        return state().get(row, col);
    }

    public HintState getRowState(int row) {
        return state().row2state[row];
    }

    public HintState getColState(int col) {
        return state().col2state[col];
    }

    public HintState getPosState(Position p) {
        return state().pos2state.get(p);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/futoshiki/domain/FutoshikiGame.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.futoshiki.domain

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.Position

class FutoshikiGame(layout: List<String>, gi: GameInterface<FutoshikiGame, FutoshikiGameMove, FutoshikiGameState>, gdi: GameDocumentInterface) : CellsGame<FutoshikiGame, FutoshikiGameMove, FutoshikiGameState>(gi, gdi) {
    companion object {
        var offset = arrayOf(
            Position(-1, 0),
            Position(0, 1),
            Position(1, 0),
            Position(0, -1)
        )
    }

    var objArray: CharArray
    var pos2hint = mutableMapOf<Position, Char>()

    operator fun get(row: Int, col: Int) = objArray[row * cols() + col]
    operator fun get(p: Position) = this[p.row, p.col]
    operator fun set(row: Int, col: Int, obj: Char) {objArray[row * cols() + col] = obj}
    operator fun set(p: Position, obj: Char) {this[p.row, p.col] = obj}

    init {
        size = Position(layout.size, layout[0].length)
        objArray = CharArray(rows() * cols())
        for (r in 0 until rows()) {
            val str = layout[r]
            for (c in 0 until cols()) {
                val p = Position(r, c)
                val ch = str[c]
                this[p] = ch
                if ((r % 2 != 0 || c % 2 != 0) && ch != ' ') pos2hint[p] = ch
            }
        }
        val state = FutoshikiGameState(this)
        states.add(state)
        levelInitilized(state)
    }

    private fun changeObject(move: FutoshikiGameMove, f: (FutoshikiGameState, FutoshikiGameMove) -> Boolean): Boolean {
        if (canRedo) {
            states.subList(stateIndex + 1, states.size).clear()
            moves.subList(stateIndex, states.size).clear()
        }
        val state = cloner.deepClone(state())
        val changed = f(state, move)
        if (changed) {
            states.add(state)
            stateIndex++
            moves.add(move)
            moveAdded(move)
            levelUpdated(states[stateIndex - 1], state)
        }
        return changed
    }

    fun switchObject(move: FutoshikiGameMove) = changeObject(move, FutoshikiGameState::switchObject)
    fun setObject(move: FutoshikiGameMove) = changeObject(move, FutoshikiGameState::setObject)

    fun getObject(p: Position) = state()[p]
    fun getObject(row: Int, col: Int) = state()[row, col]
    fun getRowState(row: Int) = state().row2state[row]
    fun getColState(col: Int) = state().col2state[col]
    fun getPosState(p: Position) = state().pos2state[p]
}
